dipy/reconst/ivim.py

```diff
@@ -0,0 +1,463 @@
+        else:
+            jac = None
+
+        if fit_method == "one_stage":
+            params_in_mask, fit_stats_in_mask = one_stage(data_in_mask,
+                                                          self.gtab, x0_mask,
+                                                          jac, bounds, tol,
+                                                          options)
+        elif fit_method == "two_stage":
+            params_in_mask, fit_stats_in_mask = two_stage(data_in_mask,
+                                                          self.gtab, x0_mask,
+                                                          self.split_b, jac,
+                                                          bounds, tol,
+                                                          options)
+        else:
+            e_s = "Fit method must be either "
+            e_s += "'one_stage' or 'two_stage'"
+            raise ValueError(e_s)
+
+        if mask is None:
+            out_shape = data.shape[:-1] + (-1, )
+            ivim_params = params_in_mask.reshape(out_shape)
+            fit_stats = fit_stats_in_mask.reshape(out_shape)
+        else:
+            ivim_params = np.zeros(data.shape[:-1] + (4,))
+            ivim_params[mask, :] = params_in_mask
+            fit_stats = np.zeros(data.shape[:-1] + (4,))
+            fit_stats[mask, :] = fit_stats_in_mask
+
+        return IvimFit(self, ivim_params, fit_stats)
+
+    def predict(self, ivim_params):
+        """
+        Predict a signal for this IvimModel class instance given parameters.
+        Parameters
+        ----------
+        ivim_params : ndarray
+            The last dimension should have 4 parameters: S0, f, D_star and D
+        """
+        return ivim_function(ivim_params, self.gtab.bvals)
+
+
+class IvimFit(object):
+
+    def __init__(self, model, model_params, fit_stats):
+        """ Initialize a IvimFit class instance.
+            Parameters
+            ----------
+            The model parameters are S0, f, D_star, D
+        """
+        self.model = model
+        self.model_params = model_params
+        self.fit_stats = fit_stats
+
+    def __getitem__(self, index):
+        model_params = self.model_params
+        N = model_params.ndim
+        if type(index) is not tuple:
+            index = (index,)
+        elif len(index) >= model_params.ndim:
+            raise IndexError("IndexError: invalid index")
+        index = index + (slice(None),) * (N - len(index))
+        return type(self)(self.model, model_params[index])
+
+    @property
+    def shape(self):
+        return self.model_params.shape[:-1]
+
+    def predict(self, gtab, step=None, S0=1.0):
+        r"""
+        Given a model fit, predict the signal.
+
+        Parameters
+        ----------
+        gtab : a GradientTable class instance
+            This encodes the directions for which a prediction is made
+        """
+        return ivim_function(self.model_params, gtab.bvals)
```

Comment:
Looks like the `S0` key-word argument is not used here (and it's also not documented).

Same for the `step` kwarg -- I believe that one really doesn't need to be here (was somehow copied over from DTI).